### PR TITLE
#184347032 list all products

### PR DESCRIPTION
**Pull request by fabmukunzi**
### What does this PR do?

Allows a user to retrieve all products

### Description of Task to be completed?

Check if a user is a seller or admin and give him only products in his collection
Otherwise display all products to the user

### How should this be manually tested?

1. Clone [repo](https://github.com/atlp-rwanda/vikings-ec-bn.git)
2. Change into the directory **vikings-ec-bn**
3. Checkout to this branch: `feature/list-all-products-184347032`
4. Run the command `npm install` to install packages
5. Run `npm run reset:db`
6.Run `npm run dev`
7.In postman create a get request `localhost:5000/api/v1/user/product`

### What are the relevant pivotal tracker stories?
[#184347032](https://www.pivotaltracker.com/story/show/184347032)

### CODE REVIEW

File: src/controllers/product.controller.js [hunk 1]
```diff
@@ -13,11 +13,20 @@ export const createProduct = async (req, res) => {
...
+};
+
+export const getAllProducts = async (req, res) => {
+  const products = await ProductService.getAllProducts();
+  const { role, id } = req.user;
+  if (role === 'seller' || role === 'admin') {
+    const userProducts=products.filter(product=>(product.userId==id))
+    res.status(200).json({ products: userProducts });
+  } else {
+    res.status(200).json({products:products})
```
Comment (TuyizeeAnastase) [inline src/controllers/product.controller.js]: - use try and catch block
- and send only one successful response

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

**Review by irakozetony — APPROVED** [review]

File: src/controllers/product.controller.js [hunk 2]
```diff
@@ -12,12 +12,27 @@ export const createProduct = async (req, res) => {
...
+export const getAllProducts = async (req, res) => {
+  try {
+    const products = await ProductService.getAllProducts();
+    const { role, id } = req.user;
+    if (role === 'seller' || role === 'admin') {
+      const userProducts = products.filter((product) => product.userId == id);
+      return res.status(200).json({ products: userProducts });
+    } else {
+      return res.status(200).json({ products: products });
+    }
```
Comment (TuyizeeAnastase) [inline src/controllers/product.controller.js]: send only one response to refer to this:

`let products = await ProductService.getAllProducts();
if(condition){
    products=products.filter((product) => product.userId == id);
  }
return res.status(200).json({
   //send products in json here
})`

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

**Review by ndatumuremyi — APPROVED** [review]

**Review by TuyizeeAnastase — APPROVED** [review]

File: src/services/product.service.js [hunk 3]
```diff
@@ -4,4 +4,7 @@ export class ProductService {
   static async createProduct(product) {
     return await Products.create(product);
   }
+  static async getAllProducts(paginate) {
+    return await Products.findAll({ limit: paginate });
+  }
```
Comment (TuyizeeAnastase) [inline src/services/product.service.js]: use include to get all related entities to products

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

File: src/services/product.service.js [hunk 4]
```diff
@@ -1,7 +1,40 @@
...
+              'gender',
+            ],
+          },
+        },
+        { model: Categories ,
+        attributes:{
+          exclude:[
+            "updatedAt",
+            "createdAt"
+          ]
```
Comment (TuyizeeAnastase) [inline src/services/product.service.js]: Nothing to exclude here in category

File: src/services/product.service.js [hunk 5]
```diff
@@ -1,7 +1,40 @@
...
+              'password',
+              'updatedAt',
+              'createdAt',
+              'authCode',
+              'verified',
+              'birthdate',
+              'description',
+              'isActive',
+              'role',
+              'gender',
```
Comment (TuyizeeAnastase) [inline src/services/product.service.js]: exclude only password and authcode

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

File: src/controllers/product.controller.js [hunk 6]
```diff
@@ -12,12 +12,26 @@ export const createProduct = async (req, res) => {
...
+    return res.status(500).json({
+      error: err.message,
+      message: 'Error occured while creating a product',
+    });
+  }
+};
+
+export const getAllProducts = async (req, res) => {
+  try {
+    const { paginate } = req.query;
```
Comment (TuyizeeAnastase) [inline src/controllers/product.controller.js]: check what is limit and offset in pagination and use them here.

File: src/services/product.service.js [hunk 7]
```diff
@@ -1,7 +1,26 @@
...
+            exclude: [
+              'password',
+              'authCode',
+            ],
+          },
+        },
+        { model: Categories},
+      ],
+    });
+  }
```
Comment (TuyizeeAnastase) [inline src/services/product.service.js]: check how findAllAndCount() function work for pagination

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

**Review by Aline-Uwera — APPROVED** [review]

**Review by Abdden — APPROVED** [review]

**Review by TuyizeeAnastase — APPROVED** [review]

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**